// @codyduong/ytmusicapi/dist/parsers/explore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseRanking = exports.parseChartTrending = exports.parseChartArtist = exports.parseChartSong = void 0;
const _1 = require(".");
const songs_1 = require("./songs");
const utils_1 = require("./utils");
const TRENDS = {
    ARROW_DROP_UP: 'up',
    ARROW_DROP_DOWN: 'down',
    ARROW_CHART_NEUTRAL: 'neutral',
};
function parseChartSong(data) {
    const flex_0 = (0, utils_1.getFlexColumnItem)(data, 0);
    let parsed = {
        title: (0, utils_1.nav)(flex_0, _1.TEXT_RUN_TEXT),
        videoId: (0, utils_1.nav)(flex_0, [..._1.TEXT_RUN, ..._1.NAVIGATION_VIDEO_ID], null),
        artists: (0, songs_1.parseSongArtists)(data, 1),
        thumbnails: (0, utils_1.nav)(data, _1.THUMBNAILS),
        isExplicit: (0, utils_1.nav)(data, _1.BADGE_LABEL, null) != null,
    };
    const flex_2 = (0, utils_1.getFlexColumnItem)(data, 2);
    if (flex_2 && 'navigationEndpoint' in (0, utils_1.nav)(flex_2, _1.TEXT_RUN)) {
        parsed['album'] = {
            name: (0, utils_1.nav)(flex_2, _1.TEXT_RUN_TEXT),
            id: (0, utils_1.nav)(flex_2, [..._1.TEXT_RUN, ..._1.NAVIGATION_BROWSE_ID]),
        };
    }
    else {
        const flex_1 = (0, utils_1.getFlexColumnItem)(data, 1);
        //const _ = nav(flex_1, ['text', 'runs']);
        //@ts-expect-error: TS doesn't discriminate flow correctly here?
        parsed['views'] = (0, utils_1.nav)(flex_1, ['text', 'runs', -1, 'text']).split(' ')[0];
    }
    parsed = { ...parsed, ...parseRanking(data) };
    return parsed;
}
exports.parseChartSong = parseChartSong;
function parseChartArtist(data) {
    const flexItem = (0, utils_1.getFlexColumnItem)(data, 1);
    let subscribers = null;
    if (subscribers) {
        subscribers = (0, utils_1.nav)(flexItem, _1.TEXT_RUN_TEXT).split(' ')[0];
    }
    let parsed = {
        title: (0, utils_1.nav)((0, utils_1.getFlexColumnItem)(data, 0), _1.TEXT_RUN_TEXT),
        browseId: (0, utils_1.nav)(data, _1.NAVIGATION_BROWSE_ID),
        subscribers: subscribers,
        thumbnails: (0, utils_1.nav)(data, _1.THUMBNAILS),
    };
    parsed = { ...parsed, ...parseRanking(data) };
    return parsed;
}
exports.parseChartArtist = parseChartArtist;
function parseChartTrending(data) {
    if (data) {
        const flex_0 = (0, utils_1.getFlexColumnItem)(data, 0);
        const artists = (0, songs_1.parseSongArtists)(data, 1);
        if (artists) {
            const index = (0, utils_1.getDotSeperatorIndex)(artists);
            // last item is views for some reason
            const views = index == artists.length ? null : artists.pop()?.['name'].split(' ')[0];
            return {
                title: (0, utils_1.nav)(flex_0, _1.TEXT_RUN_TEXT),
                videoId: (0, utils_1.nav)(flex_0, [..._1.TEXT_RUN, ..._1.NAVIGATION_VIDEO_ID], null),
                playlistId: (0, utils_1.nav)(flex_0, [..._1.TEXT_RUN, ..._1.NAVIGATION_PLAYLIST_ID], null),
                artists: artists,
                thumbnails: (0, utils_1.nav)(data, _1.THUMBNAILS),
                views: views,
            };
        }
    }
    return null;
}
exports.parseChartTrending = parseChartTrending;
function parseRanking(data) {
    return {
        rank: (0, utils_1.nav)(data, [
            'customIndexColumn',
            'musicCustomIndexColumnRenderer',
            ..._1.TEXT_RUN_TEXT,
        ]),
        trend: TRENDS[(0, utils_1.nav)(data, [
            'customIndexColumn',
            'musicCustomIndexColumnRenderer',
            'icon',
            'iconType',
        ])],
    };
}
exports.parseRanking = parseRanking;
